quinn-proto/src/config.rs

```diff
@@ -160,16 +161,34 @@ impl TransportConfig {
     /// (see [`TransportConfig::mtu_discovery_config`]).
     ///
     /// Must be at least 1200, which is the default, and known to be safe for typical internet
-    /// applications. Larger values are more efficient, but increase the risk of unpredictable
-    /// catastrophic packet loss due to exceeding the network path's IP MTU. If the provided value
-    /// is higher than what the network path actually supports, packet loss will eventually trigger
-    /// black hole detection and bring it down to 1200.
-    ///
-    /// Real-world MTUs can vary according to ISP, VPN, and properties of intermediate network links
-    /// outside of either endpoint's control. Caution should be used when raising this value for
-    /// connections outside of private networks where these factors are fully controlled.
+    /// applications. Larger values are more efficient, but increase the risk of packet loss due to
+    /// exceeding the network path's IP MTU. If the provided value is higher than what the network
+    /// path actually supports, packet loss will eventually trigger black hole detection and bring
+    /// it down to [`TransportConfig::base_udp_payload_size`].
     pub fn initial_max_udp_payload_size(&mut self, value: u16) -> &mut Self {
         self.initial_max_udp_payload_size = value.max(INITIAL_MAX_UDP_PAYLOAD_SIZE);
+        assert!(self.initial_max_udp_payload_size >= self.base_udp_payload_size);
+        self
+    }
+
+    /// The UDP payload size guaranteed to be supported by the network.
+    ///
+    /// Must be at least 1200, which is the default, and lower than or equal to
+    /// [`TransportConfig::initial_max_udp_payload_size`].
+    ///
+    /// Real-world MTUs can vary according to ISP, VPN, and properties of intermediate network links
+    /// outside of either endpoint's control. Extreme care should be used when raising this value
+    /// outside of private networks where these factors are fully controlled. If the provided value
+    /// is higher than what the network path actually supports, the result will be unpredictable and
+    /// catastrophic packet loss, without a possibility of repair. Prefer
+    /// [`TransportConfig::initial_max_udp_payload_size`] together with
+    /// [`TransportConfig::mtu_discovery_config`] to set a maximum UDP payload size that robustly
+    /// adapts to the network.
+    pub fn base_udp_payload_size(&mut self, value: u16) -> &mut Self {
+        self.base_udp_payload_size = value.max(BASE_UDP_PAYLOAD_SIZE);
+        self.initial_max_udp_payload_size = self
+            .initial_max_udp_payload_size
+            .max(self.base_udp_payload_size);
```

Comment:
I'm not super happy with how this makes the builder ordering-sensitive. It's not actually harmful if the invariant here is violated, is it? Or, maybe we should clamp the effective value when consuming the config?

Reply:
I agree that making the builder ordering-sensitive is undesirable. In this particular case, raising the `initial_max_udp_payload_size` is partly a matter of convenience (if a higher payload size is guaranteed, we might as well use it from the beginning), partly of consistency (if a black whole is detected using an MTU of `initial_max_udp_payload_size`, the MTU will be reset to `max_guaranteed_udp_payload_size`, which might actualy be higher).

Short-term, clamping the value when consuming the config would get rid of the ordering issue. The only drawback I see is that we are opening the door for future confusion, when working on Quinn internals, because now we have to remember that `initial_max_udp_payload_size` must be clamped at the start by `guaranteed_max_udp_payload_size`, every time it is used.

Long-term, I would prefer a less conservative approach, replacing the current config struct by a builder which validates the config and generates a ready-to-consume config struct (either implicitly correcting user errors, as we do now, or rejecting them). The resulting struct can then be used without worrying about correcting user mistakes.

Since this PR is blocking the 0.10 release, let's go with the short-term solution for now. However, I'd be interested in hearing your thoughts on the idea of a builder.

Reply:
> partly a matter of convenience [...], partly of consistency

Wouldn't be the first case of weird behavior being possible if you supply a weird config, at least. To some extent that's a fact of rich configuration.

> now we have to remember that initial_max_udp_payload_size must be clamped at the start by guaranteed_max_udp_payload_size, every time it is used.

I think this could be mitigated by using a getter method, and removing `pub(crate)` from the fields to enforce its use, if we want to not just let weird configs be weird.

> I'd be interested in hearing your thoughts on the idea of a builder.

I'm not categorically opposed, but I'm not sure this alone is compelling motivation for one. I prefer detecting errors per-setter as it reduces the distance between an error occurring and being reported, and so far we've gotten away without many complex invariants spanning multiple fields.

On the other hand, maybe a builder provide opportunity clean up some of the other awkwardness, e.g. with difficulty constructing platform-specific defaults. I know djc leaned into builders really hard in rustls not long ago to allow their config structures to be built up more gracefully. Config API usability improvements are definitely an area worth exploring in the future.

Reply:
Thanks! I ended up keeping the `pub(crate)` for the field, because our current testing code assumes all fields are accessible, but I added a `pub(crate)` getter for code reuse.

@djc any thoughts on the idea of having a proper builder?

Reply:
It feels like it would be a PITA to have a builder for all of `TransportConfig` if we only need it for these two MTU discovery-related values. Can we isolate the MTU things into a separate type (should we move them into `MtuDiscoveryConfig`, maybe while renaming it) that we can use a builder for?

> I'm not categorically opposed, but I'm not sure this alone is compelling motivation for one. I prefer detecting errors per-setter as it reduces the distance between an error occurring and being reported, and so far we've gotten away without many complex invariants spanning multiple fields.

Definitely agree that this can be downside of builder patterns. The most important reason for rustls to use a builder pattern was to force the caller to set some values, which IIRC is not important for our use cases (in fact, we'd probably prefer for most users not to touch the `guaranteed_max_udp_payload_size`).

> On the other hand, maybe a builder provide opportunity clean up some of the other awkwardness, e.g. with difficulty constructing platform-specific defaults.

Can you explain what kind of awkwardness we have going on here?

Reply:
> should we move them into MtuDiscoveryConfig

I think this is a poor fit because `MtuDiscoveryConfig` is omitted entirely when MTUD is disabled, but the initial and minimum MTU are still significant in that case as black hole detection runs regardless.

> Can you explain what kind of awkwardness we have going on here?

MTUD is the worst of it, since we want it on by default iff the underlying socket supports disabling fragmentation, but `TransportConfig::default` doesn't know what backend it will be used with. Maybe we should just default it on globally since as of #1547 almost every modern platform should be covered?

Segmentation offload would be similar, except there's probably no need to let users tune it, so it can remain specified via sidechannels. Maybe MTUD should require a config *and* a sidechannel DONTFRAG-compliance flag...

Reply:
I've now implemented that last strategy in #1547; I'm pretty happy with it.